feat: add name and token_last_characters to apitoken

mdtro opened the pull request:
- Add a nullable `name` column to the `ApiToken` model to help users identify and give meaningful names to their API tokens.
- Add a nullable `last_token_characters` column that will contain the last four characters of the API token to help users with identify the token in the UI when the majority of it is obfuscated.

Supports https://github.com/getsentry/sentry/issues/58918

mdtro committed:
feat: add name and token_last_characters to apitoken

mdtro committed:
add test for migration

wedamija commented on src/sentry/models/apitoken.py:
```diff
@@ -53,6 +55,15 @@ class Meta:
+        # and save them in the `token_last_characters` field so users can identify
+        # tokens in the UI where they're mostly obfuscated
+        token_last_characters = self.token[-4:]
+        self.token_last_characters = token_last_characters
```
Is there a reason we want to store this in the database, vs just obfuscating what we return from the api?

mdtro replied:
The eventual goal is to not store these tokens in plaintext, but rather as a SHA256 hash.

mdtro committed:
typing: fix save method signature

mdtro committed:
comment save function

save() is needed after the migration has been applied, not along with
otherwise we cannot properly test it and the backfill

azaslavsky commented on src/sentry/models/apitoken.py:
On L102, we have some logic for what we do if, at bulk import time, we have token collisions (which is guaranteed to happen if, for example, a user is imported multiple times). Currently, we just give new tokens to the incoming model to avoid the collision.

2 things we should look at here:
- Make sure to update `token_last_characters` at the same time we generate new tokens (may even be worth making this a Postgres level constraint, lest future developers forget to make this change in tandem somewhere).
- Decide what to do with the name - should we rename tokens when they are changed in this manner (<- we discussed this offline a bit).

mdtro replied:
I updated the backup tests to  use a value for the `name`. 

I also removed specifying the `token` field. UUID4 doesn't have any secure cryptographic guarantees, but the `secrets` module that we use for the `generate_token()` function as the default on the model does.

For the `token_last_characters`, my plan is to set these by overriding the `save()` method on the model. That will be done in the next PR. I'll add appropriate tests along with it.

azaslavsky reviewed (commented):
Do you mind updating the code in `tests/sentry/backup/test_models.py` and `src/sentry/testutils/helpers/backups.py` (<- 2 usages) to use these fields as well?

Also, more generally, may be worthwhile adding a couple of general-case tests in ex: test_authetnication.py, test_sentryapp.py, etc.

mdtro committed:
tests: update backup tests to use name on ApiToken

mdtro committed:
include token_last_characters in apitoken backup comparator

mdtro committed:
tests: accept backup snapshot changes

mdtro commented on src/sentry/migrations/0584_apitoken_add_name_and_last_chars.py:
```diff
@@ -0,0 +1,36 @@
+    # - Adding indexes to large tables. Since this can take a long time, we'd generally prefer to
+    #   have ops run this and not block the deploy. Note that while adding an index is a schema
+    #   change, it's completely safe to run the operation after the code has deployed.
+    is_dangerous = False
```
@wedamija Since I'm just adding two nullable columns with this migration, it is safe to keep this set as `is_dangerous = False`, correct?
